### PR TITLE
Added code for Linear and Binary Search

### PR DESCRIPTION
**Pull request by GSNCodes**
Added the code for linear and binary search in python #23

### CODE REVIEW

File: dsalgo/Binary_Search.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Returns index of x in the list/array if present, else -1 
+def binarySearch(arr, l, r, x):
```
Comment (codesankalp) [inline dsalgo/Binary_Search.py:2]: use oop.
add docstrings for each function.
you have to write option for algorithm in the class

File: dsalgo/Binary_Search.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            return binarySearch(arr, l, mid-1, x) 
+
+
+        else: 
+            return binarySearch(arr, mid + 1, r, x) 
+
+    else:  
+        return -1
+
+if __name__ == '__main__':
```
Comment (codesankalp) [inline dsalgo/Binary_Search.py:22]: don't write test cases here. use a separate file instead

**Review by codesankalp — CHANGES_REQUESTED** [review]
Do the requested changes.

**Comment by GSNCodes** [comment]
@codesankalp 
Have addressed your comments on the PR. Followed your example on Sorting. Please review and let me know :)

**Comment by Aryamanz29** [comment]
> @codesankalp 
> Have addressed your comments on the PR. Followed your example on Sorting. Please review and let me know :)

Add tests using python unit test framework: unittest

**Comment by GSNCodes** [comment]
@Aryamanz29 
I have added the testcases using unittest framework. Let me know if they are ok :)

**Comment by codesankalp** [comment]
@GSNCodes according to the issue you have to implement all searching algorithms to make this pull request ready to merge.

File: dsalgo/Search.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+		self.element = element
+
+		if self.algorithm == 'linear':
+			return linear_search(self.array, self.element)
+
+		elif self.algorithm == 'binary':
+			return binary_search(self.array, 0, len(self.array)-1, self.element)
+
+		else:
+			stderr.write("Error: unsupported searching algorithm passed!")
```
Comment (codesankalp) [inline dsalgo/Search.py:34]: if you have tested your code properly then you must understand the error in this line.
Don't do contributions for pull request do contributions for collaboration.

File: dsalgo/Search.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+	for index, value in enumerate(array):
+		if value == element:
+			return index
+
+	return -1
+
+
+
+def binary_search(array, l, r, element):
```
Comment (codesankalp) [inline dsalgo/Search.py:57]: code for binary search is wrong. you need to pass sorted list here.

File: tests/search_tests.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import unittest
+from dsalgo.Search import Search
+array = [2, 4, 6, 8, 10, 1]
+
```
Comment (codesankalp) [inline tests/search_tests.py:4]: write proper test cases

**Review by codesankalp — CHANGES_REQUESTED** [review]

File: dsalgo/Search.py (hunk 3 above)
Comment (GSNCodes) [reply dsalgo/Search.py:34]: I have missed the "sys" import when I did the commit. I'll correct it. 
I am doing it for collaboration and that's the reason I am doing it before October. :)

File: dsalgo/Search.py (hunk 4 above)
Comment (GSNCodes) [reply dsalgo/Search.py:57]: Tell me how you plan to design the library. 
Do you want to add sorting to be an option ? Like should I sort it by default, for binary search or do you want me to throw an error if the array given is not sorted?

Because if we sort it, the index positions would change. Do you want me to output the index value or just return true or false corresponding to whether the element is present or not?

@codesankalp

File: tests/search_tests.py (hunk 5 above)
Comment (GSNCodes) [reply tests/search_tests.py:4]: I can cover all corner cases but you must decide upon how you want to design it. 
Could you update the details of your expected design here and in the issue section so that I can refer to that and meet your expectations.

File: dsalgo/Search.py (hunk 4 above)
Comment (codesankalp) [reply dsalgo/Search.py:57]: I want you to show a warning when array is not sorted for binary search. But search the element in sorted one @GSNCodes.

**Comment by codesankalp** [comment]
I am opening this pull request again because you have worked on this issue and i don't want to waste your work.
Try to implement all algo for this.
But still you want to close this pull request then mention the reason.

File: tests/search_tests.py (hunk 5 above)
Comment (codesankalp) [reply tests/search_tests.py:4]: for writing code you can view the linked list implementation.
You don't need to follow a proper template only thing that we expect you from is modular code.
Regarding test cases you have to think of all edge cases by your own.
Designing is your part but be concise of what you are doing.
